STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,199 @@
+  if (registry_editor[0].used_flag == REGISTRY_FREE)
+  {
+    flash_ptr = (uint8_t*)(&__REGISTRY);
+  }
+  else
+  {
+    az_registry_block last_block;
+
+    // The beginning of buffer space
+    uint8_t* max_address = (uint8_t*)(&__REGISTRY);
+
+    // registry with content
+    for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+    {
+      if (registry_editor[i].used_flag == REGISTRY_USED)
+      {
+        if (az_span_ptr(registry_editor[i].value) > max_address)
+        {
+          max_address = az_span_ptr(registry_editor[i].value);
+          last_block = registry_editor[i];
+        }
+      }
+    }
+    // last_block should have reference to the last block in available memory, rounded address
+    int32_t value_buf_size = ((((az_span_size(last_block.value) - 1) >> 3) + 1) << 3);
+    flash_ptr = az_span_ptr(last_block.value) + value_buf_size;
+  }
+
+  return flash_ptr;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_get(az_span key, az_span* value)
+{
+  az_result result;
+
+  // registry with content
+  for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+  {
+    if (registry_editor[i].used_flag == REGISTRY_USED)
+    {
+      if (az_span_is_content_equal(key, registry_editor[i].key))
+      {
+        *value = registry_editor[i].value;
+        result = AZ_OK;
+        break;
+      }
+    }
+    else
+    {
+      result = AZ_ERROR_ITEM_NOT_FOUND;
+      break;
+    }
+  }
+
+  return result;
+}
+
+AZ_NODISCARD az_result az_ulib_registry_add(az_span new_key, az_span new_value)
+{
+  AZ_ULIB_TRY
+  {
+    int32_t free_block_index;
+    HAL_StatusTypeDef hal_status;
+    az_span sample_value;
+    az_result result;
+
+    // validate for duplicates before adding new entry
+    result = az_ulib_registry_get(new_key, &sample_value);
+    AZ_ULIB_THROW_IF_ERROR((result == AZ_ERROR_ITEM_NOT_FOUND), AZ_ERROR_ULIB_ELEMENT_DUPLICATE);
+
+    // look for next available spot in list
+    for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+    {
+      if (registry_editor[i].used_flag == REGISTRY_FREE)
+      {
+        free_block_index = i;
+        break;
+      }
+    }
+
+    // find destination in flash buffer
+    uint8_t* key_dest_ptr = flash_buf_get();
+    uint8_t* val_dest_ptr = key_dest_ptr + ((((az_span_size(new_key) - 1) >> 3) + 1) << 3);
+
+    // set free block information
+    registry_editor[free_block_index].used_flag = REGISTRY_USED;
+    registry_editor[free_block_index].key = az_span_create(key_dest_ptr, az_span_size(new_key));
+    registry_editor[free_block_index].value = az_span_create(val_dest_ptr, az_span_size(new_value));
+
+    // write key to flash
+    if ((hal_status
+         = internal_flash_write(key_dest_ptr, az_span_ptr(new_key), az_span_size(new_key)))
+        != HAL_OK)
+    {
+      result = result_from_hal_status(hal_status);
+    }
+
+    // write value to flash
+    if ((hal_status
+         = internal_flash_write(val_dest_ptr, az_span_ptr(new_value), az_span_size(new_value)))
```

Comment:
>internal_flash_write [](http://example.com/codeflow?start=11&length=20)

What happen if this operation fail? You will have a valid Key with a invalid value in the memory. Or what if the device shutdown in the meddle of the write process?